runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/StreamingDataflowWorker.java

```diff
@@ -1900,16 +1903,47 @@ private void sendWorkerUpdatesToDataflowService(
           // WorkItem.
           if (item.getCumulative()) {
             item.setCumulative(false);
+            // Group counterUpdates by counterUpdateKey so they can be aggregated before sending to
+            // dataflow service.
+            fnApiCounters
+                .computeIfAbsent(getCounterUpdateKey(item), k -> new ArrayList<>())
+                .add(item);
           } else {
             // In current world all counters coming from FnAPI are cumulative.
             // This is a safety check in case new counter type appears in FnAPI.
             throw new UnsupportedOperationException(
                 "FnApi counters are expected to provide cumulative values."
-                    + " Please, update convertion to delta logic"
+                    + " Please, update conversion to delta logic"
                     + " if non-cumulative counter type is required.");
           }
+        }
 
-          counterUpdates.add(item);
+        // Aggregates counterUpdates with same counterUpdateKey to single CounterUpdate if possible
+        // so we can avoid excessive I/Os for reporting to dataflow service.
+        List<CounterUpdateAggregator> availableAggregators =
+            CounterUpdateAggregator.getAllAvailableCounterUpdateAggregators();
+        for (List<CounterUpdate> counterUpdateList : fnApiCounters.values()) {
+          if (counterUpdateList.size() == 0) {
+            continue;
+          }
+          CounterUpdate head = counterUpdateList.get(0);
+          boolean aggregatable = false;
+          for (CounterUpdateAggregator aggregator : availableAggregators) {
+            if (aggregator.isCorrespondingCounterUpdate(head)) {
+              counterUpdates.add(aggregator.aggregate(counterUpdateList));
+              aggregatable = true;
+              break;
+            }
+          }
+          if (!aggregatable) {
+            LOG.debug(
```

Comment:
Lets add a debug log if we are left with any counters for a key which is not aggregated and more than 10 values of a single key is present.

Reply:
We can keep this log as it is. And the aggregated log can be an warning log.

Reply:
It would also be worth to limit period when we report this log to avoid spam.

Reply:
done.